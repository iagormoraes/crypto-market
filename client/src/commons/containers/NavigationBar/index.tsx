import { useCallback } from 'react';
import Image from 'next/image';
import Link from 'next/link';
import { useRouter } from 'next/router';
import { signOut } from 'next-auth/react';
import styles from './style.module.css';

import Button, { ButtonVariant } from '../../components/buttons/Button';

import useAppSession, { AppSessionState } from '../../../hooks/useAppSession';
import { UserRolesEnum } from '../../../modules/user/interfaces/user-roles.interface';

export default function NavigationBar() {
  const { push } = useRouter();
  const { status, session } = useAppSession();

  const renderOptions = useCallback(() => {
    switch (status) {
      case AppSessionState.Loading:
        return null;
      case AppSessionState.Unauthenticated:
        return (
          <>
            <li>
              <Link href="/login" passHref>
                <a>
                  <Button variant={ButtonVariant.Small}>Login</Button>
                </a>
              </Link>
            </li>
            <li>
              <Link href="/register" passHref>
                <a>
                  <Button variant={ButtonVariant.Small} alt>
                    Register
                  </Button>
                </a>
              </Link>
            </li>
          </>
        );

      case AppSessionState.Authenticated:
        return (
          <>
            {session.user?.role === UserRolesEnum.Admin && (
              <li>
                <Link href="/admin/dashboard" passHref>
                  <a>
                    <Button variant={ButtonVariant.Small} alt>
                      Admin
                    </Button>
                  </a>
                </Link>
              </li>
            )}
            <li>
              <Button
                variant={ButtonVariant.Small}
                alt
                onClick={() => {
                  push('/');
                  signOut({ redirect: false });
                }}
              >
                Logout
              </Button>
            </li>
          </>
        );

      default:
        return null;
    }
  }, [status, session, push]);

  return (
    <nav className={styles.nav}>
      <div className="container m-auto flex items-center px-2">
        <Link href="/" shallow={false}>
          <a className="flex items-center">
            <div className={styles.logo}>
              <Image
                className="w-full h-full"
                src="/logo.svg"
                alt="Crypto Market logo"
                layout="fill"
              />
            </div>
            <h1 className="ml-2 text-white text-lg font-bold hidden md:block">
              Crypto Market
            </h1>
          </a>
        </Link>
        <nav className="ml-auto">
          <ul className="flex items-center gap-4">
            {session?.user?.name && (
              <span className="text-white mr-4 invisible md:visible">
                Hello, <strong>{session.user.name}</strong>
              </span>
            )}
            {renderOptions()}
          </ul>
        </nav>
      </div>
    </nav>
  );
}
