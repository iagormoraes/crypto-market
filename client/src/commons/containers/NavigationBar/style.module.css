.nav {
    min-height: 1rem;

    @apply bg-blueGray-900 bg-opacity-60 backdrop-blur-md w-full p-2 z-30;
    @apply absolute top-0;
}

.logo {
    @apply w-12 h-12 relative fill-white;
}
