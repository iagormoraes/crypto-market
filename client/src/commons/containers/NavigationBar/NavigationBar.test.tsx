import { render } from '@testing-library/react';
import NavigationBar from '.';
import { SessionProvider } from 'next-auth/react';

global.fetch = jest.fn(
  (): Promise<any> =>
    Promise.resolve({
      json: () => Promise.resolve({}),
    }),
);

jest.mock('next/router', () => ({
  useRouter() {
    return {
      route: '/',
      pathname: '',
      query: '',
      asPath: '',
      push: jest.fn(),
    };
  },
}));

describe('<NavigationBar/>', () => {
  it('matches snapshot on loading', () => {
    const { container } = render(<NavigationBar />, {
      wrapper: ({ children }) => <SessionProvider>{children}</SessionProvider>,
    });

    expect(container).toMatchSnapshot();
  });
});
