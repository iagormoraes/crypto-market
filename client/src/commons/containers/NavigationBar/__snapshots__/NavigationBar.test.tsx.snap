// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NavigationBar/> matches snapshot on loading 1`] = `
<div>
  <nav
    class="nav"
  >
    <div
      class="container m-auto flex items-center px-2"
    >
      <a
        class="flex items-center"
        href="/"
      >
        <div
          class="logo"
        >
          <span
            style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
          >
            <img
              alt="Crypto Market logo"
              class="w-full h-full"
              data-nimg="fill"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
            />
            <noscript />
          </span>
        </div>
        <h1
          class="ml-2 text-white text-lg font-bold hidden md:block"
        >
          Crypto Market
        </h1>
      </a>
      <nav
        class="ml-auto"
      >
        <ul
          class="flex items-center gap-4"
        />
      </nav>
    </div>
  </nav>
</div>
`;
